README.md
# Markdown text

example:

```javascript
var asd = "asdfa sd";
```

Usage:
Launch the x-world.xhtml on any browser and enjoy the game. Tested on Chrome, mozilla and Internet explorer.

Please feel free to suggest improvements.


change_log.txt
Version v1_3
Author Chethan Ningaraju

Major Changes
=============

[x] Changes related to avoiding the reload of images for every canvas update. The images are loaded once and different loaded images are used for canvas update.


hbbtvlib.js
/**
 * HbbTV library v1.0
 * 
 * (C) 2009, IRT GmbH
 * 
 * Overview:
 * 
 * function hbbtvlib_intialize()
 *   Creates and initializes HbbTV, i.e. OIPF DAE, embedded objects. 
 *   Always call this function prior to other functions of this library.
 * 
 * function hbbtvlib_show()
 *   Shows the application and requests keys via the keyset object.
 *   
 * function hbbtvlib_hide()
 *   Hides the application and releases keys via the keyset object.
 *   
 * function hbbtvlib_createApp()
 *   Create applications with the OIPF Application Manager
 *   
 * function hbbtvlib_closeApp()
 *   Destroys this application.
 * 
 * function hbbtvlib_init_broadcast()
 *   Convenience function that integrates the broadcast video in your application.
 *   
 * function hbbtvlib_release_broadcast()
 *   Convenience function that removes the broadcast video from your application.
 *
 * function hbbtvlib_current_service()
 *   Returns the DVB service ID.
 */

/**
 * The last error occured in hbbtvlib. If one of the functions returns false, 
 * the error can be retrieved via this property.
 */
var hbbtvlib_lastError = null;

/**
 * @return true if OIPF functions are available
 */
function hbbtvlib_initialize() {
	try {
		int_createOipfObjs();
		
		var appMgr = int_objs[int_objTypes.appMan];
		if (appMgr && typeof(appMgr.getOwnerApplication) != "undefined")
			int_app = appMgr.getOwnerApplication(document);
		else
			int_app = null;
		
		int_initKeysetObj();
	} catch (e) {
		hbbtvlib_lastError = e;
		int_app = null;
	}
	return (int_app!=null);
};

/**
 * This function does everything HbbTV requires an application to do
 * when it wants to get visible. 
 * 
 * @return
 */
function hbbtvlib_show() {
	try {
		int_app.show();
		if (typeof int_app.activate != "undefined") int_app.activate();
		if (typeof int_app.activateInput != "undefined") int_app.activateInput();
		int_keyset.setValue(int_ksVisible);
		return true;
	} catch (e) {
		hbbtvlib_lastError = e;
		return false;
	}
}

/**
 * This function does everything HbbTV requires an application to do
 * when it wants to transition to the hidden state. 
 * 
 * @return
 */
function hbbtvlib_hide() {
	try {
		int_app.hide();
		int_keyset.setValue(int_ksHidden);
		return true;
	} catch (e) {
		hbbtvlib_lastError = e;
		return false;
	}
}

/**
 * Overrides the default keysets.
 * Default is RED for hidden state, all for visible state.
 * 
 * @param visibleSet a keyset bitmask
 * @param hiddenSet a keyset bitmask
 * @return
 */
function hbbtvlib_setKeysets(visibleSet, hiddenSet) {
	if (typeof visibleSet == "Number") int_ksVisible = visibleSet;
	if (typeof hiddenSet == "Number") int_ksHidden = hiddenSet;
};


/**
 * Starts a new application and destroys this application.
 * 
 * @param dvbUrl including organization and application ids
 * @param httpUrl a fallback URL
 * 
 * @return false
 */
function hbbtvlib_createApp(dvbUrl, httpUrl) {
	if (int_app) {
		try {
			if (dvbUrl && int_app.createApplication(dvbUrl, false)) {
				int_app.destroyApplication();
				return true;
			}
		} catch (e) {
			hbbtvlib_lastError = e;
		}
		try {
			if (httpUrl && int_app.createApplication(httpUrl, false)) {
				int_app.destroyApplication();
				return true;
			}
		} catch (e) {
			hbbtvlib_lastError = e;
		}
	} else {
		document.location.href = httpUrl;
	}	
	return false;
};

/**
 * Destroys this application.
 * 
 * @param fallBackUrl
 * @return
 */
function hbbtvlib_closeApp(fallBackUrl) {
	if (int_app) {
		try {
			int_app.destroyApplication();
		} catch (e) {
			hbbtvlib_lastError = e;
		}
	} else {
		document.location.href = fallBackUrl;
	}
	return false;
};

/**
 * Handle OIPF DAE Broadcast video
 */
var int_bc_video = null;
var int_bc_container = null;

/**
 * Creates and initializes a broadcast video inside the element
 * identified by the containerId. If no bc video can be included
 * the dummy picture is added instead of the bc.
 * 
 * @param parentId the id of the HTML container where the video/broadcast object will be added
 * @param objId id which shall be set for the video/broadcast object
 * @param dummyPic an optional picture to be shown if video/broadcast can not be added
 * 
 * @return the video/broadcast object or false 
 */
function hbbtvlib_init_broadcast (parentId, objId, dummyPic) {
	try {
		int_bc_container = document.getElementById(parentId);
		if (!int_bc_container) return false;

		// root container for video/broadcast object
		int_bc_container.innerHTML = '<object id="' + objId + '" type="video/broadcast"> </object>';
		int_bc_video = document.getElementById(objId);

		if (typeof (int_bc_video.fullScreen) != 'undefined') {
			int_bc_video.setFullScreen(false);
		}
		if (typeof (int_bc_video.bindToCurrentChannel) != 'undefined') {
			int_bc_video.bindToCurrentChannel();
		}
		
		// check if OIPF video object is supported
		if (typeof (int_bc_video.currentChannel) != 'undefined') return int_bc_video;

	} catch (e) {
		hbbtvlib_lastError = e;
	}
	if (dummyPic) int_bc_container.innerHTML = '<img id="'+ objId + '" src="' + dummyPic + '" alt="TV Bild" />';
	return false;
};

/**
 * Releases the video/broadcast object and removes the object from the DOM.
 */
function hbbtvlib_release_broadcast () {
	try {
		if (int_bc_video && typeof (int_bc_video.release) != "undefined") {
			int_bc_video.release();
			int_bc_video = null;
		}
		if (int_bc_container) {
			int_bc_container.innerHTML = "";
			int_bc_container = null;
		}
	} catch (e) {
		hbbtvlib_lastError = e;
	}
};

/**
 * @return the DVB service id of the current presented service or -1
 */
function hbbtvlib_current_service () {
	try {
		return int_bc_video.currentChannel.sid;
	} catch (e) {
		hbbtvlib_lastError = e;
	}
	return -1;
};

/*
 * Library internal (private) properties and functions.
 */

/**
 * Array of OIPF DAE embedded objects. Access Key is the mime-type without "application/".
 */
var int_objs = new Array();

var int_objTypes = {
	appMan: "oipfApplicationManager",
	config: "oipfConfiguration"
};

/**
 * OIPF DAE Application object
 * 	- to show/hide the application
 * 	- create or destroy applications
 *  - providing the keyset object
 */
var int_app = null;

/**
 * The OIPF keyset object used to request keys from terminal.
 * 
 * Usually this is the RED button in hidden mode, and up to all
 * available keys in visible mode.
 */
var int_keyset = null;

var int_ksHidden = null;

var int_ksVisible = null;

/**
 * Internal function to create OIPF embedded objects,
 * like ApplicationManager etc., if not already existing.
 */
function int_createOipfObjs() {
	if (int_objs.length) return;
	// Look for existing objects
	var objects = document.getElementsByTagName("object");
	for (var i=0; i<objects.length; i++) {
		var sType = objects.item(i).getAttribute("type");
		//alert(sType);
		for (var eType in int_objTypes) {
			if (sType == "application/" + int_objTypes[eType]) {
				int_objs[int_objTypes[eType]] = objects.item(i);
			}
		}
	}
	
	// create missing objects
	var oipfObjs = document.createElement("div");
	var objCreated = false;
	for (var typeId in int_objTypes) {
		var type = int_objTypes[typeId];
		if (!int_objs[type]) {
			var obj = document.createElement("object");
			obj.setAttribute("type", "application/"+type);
			oipfObjs.appendChild(obj);
			int_objs[type] = obj;
			objCreated = true;
		}
	}
	if (objCreated)	{
		oipfObjs.setAttribute("style", "visibility: hidden; width: 0; height: 0;");
		document.getElementsByTagName("body")[0].appendChild(oipfObjs);
	}
};

/**
 * Retrieves the keyset object for HbbTV 1.1.1 and 0.5 implementations.
 * 
 * Sets default keysets for hidden and visible modes.
 */
function int_initKeysetObj () {
	// try HbbTV 1.1.1
	try {
		int_keyset = int_app.privateData.keyset;
		hbbtvlib_lastError = "privateData";
	} catch (e) {
		int_keyset = null;
		// try HbbTV 0.5
		try {
			var tmp = int_objs[int_objTypes.config].keyset;
			tmp.setValue = function (val) {this.value = val;};
			int_keyset = tmp;
			hbbtvlib_lastError = "oipfConfig";
			
		} catch (e) {
			int_keyset = null;
		}
	}

	if (int_keyset) {
		// set default key sets
		if (!int_ksHidden)  int_ksHidden  = int_keyset.RED;
		if (!int_ksVisible) int_ksVisible =  0x33F; // color + nav + vcr + numeric + alpha
	}

	return int_keyset != null;
};


keycodes.js
/* "redbutton-fix" */
if (typeof (VK_RED) == "undefined") {
  var VK_RED      = 403;
  var VK_GREEN    = 404;
  var VK_YELLOW   = 405;
  var VK_BLUE     = 406;

  var VK_LEFT     = 37;
  var VK_UP       = 38;
  var VK_RIGHT    = 39;
  var VK_DOWN     = 40;
  var VK_ENTER    = 13;

  var VK_0        = 48;
  var VK_1        = 49;
  var VK_2        = 50;
  var VK_3        = 51;
  var VK_4        = 52;
  var VK_5        = 53;
  var VK_6        = 54;
  var VK_7        = 55;
  var VK_8        = 56;
  var VK_9        = 57;

  var VK_PLAY     = 415;
  var VK_PAUSE    = 19;
  var VK_STOP     = 413;
  var VK_FAST_FWD = 417;
  var VK_REWIND   = 412;

  var VK_BACK     = 461;

  var VK_TELETEXT = 459;
}

if (typeof(KeyEvent)!='undefined') {
  if (typeof(KeyEvent.VK_LEFT)!='undefined') {
    var VK_LEFT = KeyEvent.VK_LEFT;
    var VK_UP = KeyEvent.VK_UP;
    var VK_RIGHT = KeyEvent.VK_RIGHT;
    var VK_DOWN = KeyEvent.VK_DOWN;
  }
  if (typeof(KeyEvent.VK_ENTER)!='undefined') {
    var VK_ENTER = KeyEvent.VK_ENTER;
  }
  if (typeof(KeyEvent.VK_RED)!='undefined') {
    var VK_RED = KeyEvent.VK_RED;
    var VK_GREEN = KeyEvent.VK_GREEN;
    var VK_YELLOW = KeyEvent.VK_YELLOW;
    var VK_BLUE = KeyEvent.VK_BLUE;
  }
  if (typeof(KeyEvent.VK_PLAY)!='undefined') {
    var VK_PLAY = KeyEvent.VK_PLAY;
    var VK_PAUSE = KeyEvent.VK_PAUSE;
    var VK_STOP = KeyEvent.VK_STOP;
  }
  if (typeof(KeyEvent.VK_BACK)!='undefined') {
    var VK_BACK = KeyEvent.VK_BACK;
  }
  if (typeof(KeyEvent.VK_0)!='undefined') {
    var VK_0 = KeyEvent.VK_0;
    var VK_1 = KeyEvent.VK_1;
    var VK_2 = KeyEvent.VK_2;
    var VK_3 = KeyEvent.VK_3;
    var VK_4 = KeyEvent.VK_4;
    var VK_5 = KeyEvent.VK_5;
    var VK_6 = KeyEvent.VK_6;
    var VK_7 = KeyEvent.VK_7;
    var VK_8 = KeyEvent.VK_8;
    var VK_9 = KeyEvent.VK_9;
  }
}
if (typeof(VK_LEFT)=='undefined') {
  var VK_LEFT = 0x25;
  var VK_UP = 0x26;
  var VK_RIGHT = 0x27;
  var VK_DOWN = 0x28;
}
if (typeof(VK_ENTER)=='undefined') {
  var VK_ENTER = 0x0d;
}
if (typeof(VK_RED)=='undefined') {
  var VK_RED = 0x74;
  var VK_GREEN = 0x75;
  var VK_YELLOW = 0x76;
  var VK_BLUE = 0x77;
}
if (typeof(VK_PLAY)=='undefined') {
  var VK_PLAY = 0x50;
  var VK_PAUSE = 0x51;
  var VK_STOP = 0x53;
}
if (typeof(VK_BACK)=='undefined') {
  var VK_BACK = 0xa6;
}
if (typeof(VK_0)=='undefined') {
  var VK_0 = 0x30;
  var VK_1 = 0x31;
  var VK_2 = 0x32;
  var VK_3 = 0x33;
  var VK_4 = 0x34;
  var VK_5 = 0x35;
  var VK_6 = 0x36;
  var VK_7 = 0x37;
  var VK_8 = 0x38;
  var VK_9 = 0x39;
}


first_game.js
/*
*Author - Chethan Ningaraju
*/

window.addEventListener('load', function() {
  //hbbtvlib_initialize();
  //hbbtvlib_show();
});
window.addEventListener('keydown', function(e) {
  switch (e.keyCode) {
    case 40:
      e.preventDefault();
      if (userStartGame == 0) {
        myGameArea.interval = setInterval(updateGameArea, UpdateInterval);
        backgroundSound.play();
        userStartGame = 1;
      } else {
        document.getElementById('output').innerHTML = 'Red Key';
        myGamePiece.y += Math.round(20 * scale_factor);
        break;
      }
    case 38:
      e.preventDefault();
      document.getElementById('output').innerHTML = 'Yellow Key';
      myGamePiece.y -= Math.round(20 * scale_factor);
      myGamePiece.myimage = rocket_on;
      myGameArea.lastBottonFrame = myGameArea.frameNo
      break;
    case 27:
      e.preventDefault();
      hbbtvlib_init_broadcast('video-container', 'video');
      break;
  }
}, false);

var myGamePiece;
var myObstacles = [];
var myScore;
var gameBackground;
var backgroundSound;
var userStartGame = 0;

// two images for fire animation
var fireIm_1 = new image_component("./images/fire_5.png");
var fireIm_2 = new image_component("./images/fire_6.png");
var rocket_off = new image_component("./images/rocket_off_scaled.png");
var rocket_on = new image_component("./images/rocket_on_scaled.png");
var gameBackGroundImage = new image_component("./images/sky_background.jpg");
//back gound sound
var musicBG = "./sounds/TheForestAwakes.mp3"

//Game Parameters
var scale_factor = 1.6;// Everything pixel related should be scaled with this
var dispX = Math.round(480 * scale_factor), dispY = Math.round(270 * scale_factor);
var pieceWidth = Math.round(30 * scale_factor), pieceHeight = Math.round(30 * scale_factor);
var minGap = pieceHeight + Math.round(20 * scale_factor), maxGap = dispY * 0.8;

// FPS
var UpdateInterval = 20;
var pixChangePerFrame = Math.round(Math.round(UpdateInterval / 20 * scale_factor));
function startGame() {
  console.log('check2');
  myGameArea.start();
  gameBackground = new component(dispX, dispY, gameBackGroundImage, 0, 0, "image", "background");
  myGamePiece = new component(pieceWidth, pieceHeight, rocket_off, Math.round(10 * scale_factor), Math.round(120 * scale_factor), "image", "piece");
  myScore = new component("15px", "Consolas", "red", Math.round(380 * scale_factor), Math.round(40 * scale_factor), "text");
  backgroundSound = new sound(musicBG);
}
/**
 * @constructor
 */
function sound(src) {
  this.sound = document.createElement("audio");
  this.sound.src = src;
  this.sound.setAttribute("preload", "auto");
  this.sound.setAttribute("controls", "none");
  this.sound.style.display = "none";
  document.body.appendChild(this.sound);
  this.play = function() {
    this.sound.play();
  }
  this.stop = function() {
    this.sound.pause();
  }
}
/**
 * @constructor
 * @param {string} src
 */
function image_component(source) {
  this.image = new Image();
  this.image.src = source;
  this.isLoaded = false;
  this.image.onload = function() {
    this.isLoaded = true;
  };
}
/**
 * @constructor
 * @param {number|string} width
 * @param {string} height
 */
function component(width, height, color, x, y, type, imType) {
  this.type = type;
  this.width = width;
  this.height = height;
  this.x = x;
  this.y = y;
  this.color = color;
  if (this.type == "image") {
    this.myimage = color;
    this.imType = imType;
  }
}
component.prototype.update = function() {
  ctx = myGameArea.context;
  if (this.type == "text") {
    ctx.font = this.width + " " + this.height;
    ctx.fillStyle = this.color;
    ctx.fillText(this.text, this.x, this.y);
  }
  else if (this.type == "image") {
    if (this.imType == "background") {
      var cur_scroll_start = (this.x + Math.floor((myGameArea.frameNo + 1) / 2)) % this.myimage.image.width;
      //window.alert(this.myimage.isLoaded);
      ctx.drawImage(this.myimage.image, cur_scroll_start, this.y, this.width, this.myimage.image.height, this.x, this.y, this.width, this.height);
    }
    else {
      ctx.drawImage(this.myimage.image, this.x, this.y, this.width, this.height);
    }
    //check if the image is rocket on, It should be on only for some duration
    if (this.imType == "piece" && myGameArea.lastBottonFrame != -1 && (myGameArea.frameNo - myGameArea.lastBottonFrame) > 10) {
      this.myimage = rocket_off;
      myGameArea.lastBottonFrame = -1;
    } else if (this.imType == "obstacle") {
      if (myGameArea.frameNo != 1 && everyinterval(UpdateInterval / 2)) {
        if (this.myimage == fireIm_1) {
          this.myimage = fireIm_2;
        } else {
          this.myimage = fireIm_1;
        }
      }
    }
    else if (this.imType == "backgound") {

    }
  }
  else {
    ctx.fillStyle = this.color;
    ctx.fillRect(this.x, this.y, this.width, this.height);
  }
};
component.prototype.crashWith = function(otherobj) {
  var myleft = this.x;
  var myright = this.x + (this.width);
  var mytop = this.y;
  var mybottom = this.y + (this.height);
  var otherleft = otherobj.x;
  var otherright = otherobj.x + (otherobj.width);
  var othertop = otherobj.y;
  var otherbottom = otherobj.y + (otherobj.height);
  var crash = true;
  if ((mybottom < othertop) || (mytop > otherbottom) || (myright < otherleft) || (myleft > otherright)) {
    crash = false;
  }
  return crash;
};
function updateGamePiece() {
  if (myGamePiece.y >= myGameArea.canvas.height - myGamePiece.height) {
    myGamePiece.y = myGameArea.canvas.height - myGamePiece.height;
  }
  else {
    myGamePiece.y += pixChangePerFrame;
  }
}
function updateGameArea() {
  var x, height, gap, minHeight, maxHeight;;
  for (i = 0; i < myObstacles.length; i += 1) {
    if (myGamePiece.crashWith(myObstacles[i])) {
      myGameArea.stop();
      return;
    }
  }
  myGameArea.clear();
  myGameArea.frameNo += 1;
  if (myGameArea.frameNo == 1 || everyinterval(150)) {
    x = myGameArea.canvas.width;//this will make sure it forms at other end of frame
    minHeight = Math.round(20 * scale_factor);
    maxHeight = Math.round(200 * scale_factor);
    height = Math.floor(Math.random() * (maxHeight - minHeight + 1) + minHeight);
    gap = Math.floor(Math.random() * (maxGap - minGap + 1) + minGap);

    y = myGameArea.canvas.height - Math.round(200 * scale_factor);
    myObstacles.push(new component(Math.round(20 * scale_factor), height, fireIm_1, x, 0, "image", "obstacle"));
    myObstacles.push(new component(Math.round(20 * scale_factor), x - height - gap, fireIm_1, x, height + gap, "image", "obstacle"));
  }
  // Order is important here, based on this elements will be overwritten
  gameBackground.update();
  for (i = 0; i < myObstacles.length; i += 1) {
    myObstacles[i].x -= Math.round(1 * scale_factor);
    myObstacles[i].update();
  }
  // Write new score as frame number
  myScore.text = "SCORE:" + Math.round(myGameArea.frameNo / 10);
  myScore.update();
  updateGamePiece();
  myGamePiece.update();
}
var myGameArea = {
  canvas: document.createElement("canvas"),
  start: function() {
    this.canvas.width = dispX;
    this.canvas.height = dispY;

    this.canvas.style.zIndex = 8;
    this.canvas.style.position = "absolute";
    this.canvas.style.border = "1px solid";
    div = document.getElementById("save-area");
    this.context = this.canvas.getContext("2d");
    //document.body.insertBefore(this.canvas, document.body.childNodes[0]);
    div.appendChild(this.canvas)
    this.frameNo = 0;
    this.lastBottonFrame = -1;

    /**
     * Startup screen
     */
    var gameBackground1 = new Image();
    gameBackground1.src = "./images/start_background.jpg";
    var startMessage1 = new component("40px", "Impact", "cyan", Math.round(200 * scale_factor), Math.round(120 * scale_factor), "text");
    startMessage1.text = "X-SPACE";

    var startMessage2 = new component("20px", "Impact", "yellow", Math.round(200 * scale_factor), Math.round(140 * scale_factor), "text");
    startMessage2.text = "Use Up & Down arrows to play";

    gameBackground1.onload = function() {
      myGameArea.context.drawImage(gameBackground1, 0, 0, dispX, dispY);
      startMessage1.update();
      startMessage2.update();
    };
    //End of start up screen

  },
  clear: function() {
    this.context.clearRect(0, 0, this.canvas.width, this.canvas.height);
  },
  stop: function() {
    // Display gameover message before restarting the game
    var stopMessage = new component("30px", "Consolas", "red", Math.round(200 * scale_factor), Math.round(120 * scale_factor), "text");
    stopMessage.text = "GAME OVER";
    stopMessage.update();
    var stopMessage_2 = new component("30px", "Consolas", "red", Math.round(210 * scale_factor), Math.round(150 * scale_factor), "text");
    stopMessage_2.text = "SCORE:" + Math.round(myGameArea.frameNo / 10);
    stopMessage_2.update();
    clearInterval(this.interval);
    setTimeout(function() {
      window.location.href = window.location.href.split('?')[0] + "?t=" + (new Date().getTime());
    }, 5000)

  }
}
function everyinterval(n) {
  if ((myGameArea.frameNo / n) % 1 == 0) { return true; }
  return false;
}

setTimeout(function() {
  //delay for images to pre-load
}, 1000)
startGame();